axelrod/strategies/finite_state_machines.py

```diff
@@ -0,0 +1,277 @@
+from axelrod import Actions, Player, Game, init_args
+
+C, D = Actions.C, Actions.D
+
+
+class SimpleFSM(object):
+    """Simple implementation of a finite state machine that transitions
+    between states based on the last round of play."""
+
+    def __init__(self, transitions, initial_state):
+        """
+        transitions is a list of the form
+        [(state, last_opponent_action, next_state, next_action), ...]
+
+        TitForTat would be represented with the following table:
+        [(1, C, 1, C), (1, D, 1, D)]
+        with initial play C and initial state 1.
+
+        """
+        self.state = initial_state
+        self.state_transitions = dict()
+        for (state, opp_action, next_state, next_action) in transitions:
+            self.state_transitions[(state, opp_action)] = (next_state, next_action)
+
+    def move(self, opponent_action):
+        """Computes the response move and changes state."""
+        next_state, next_action = self.state_transitions[(self.state, opponent_action)]
+        self.state = next_state
+        return next_action
+
+
+class FSMPlayer(Player):
+    """Abstract base class for finite state machine players."""
+
+    name = "FSM Player"
+
+    classifier = {
+        'memory_depth': float('inf'),  # Long memory
+        'stochastic': False,
+        'makes_use_of': set(),
+        'inspects_source': False,
+        'manipulates_source': False,
+        'manipulates_state': False
+    }
+
+    def __init__(self, transitions, initial_state, initial_action):
+        Player.__init__(self)
+        self.initial_action = initial_action
+        self.fsm = SimpleFSM(transitions, initial_state)
+
+    def strategy(self, opponent):
```

Comment:
It's possible that an `FSMPlayer` be ill defined. If the `transitions` are not complete:

For example:

```
transitions = ((1, D, 2, D),
                      (1, C, 1, D),
                      (2, C, 1, D),
                      (2, D, 3, C),
                      (3, C, 3, C))
```

Could we either:

- Include a check for this in the strategy
- Include a generic test in the strategy `TestFSMPlayer` class that will test that all states are in the transitions. Something like checking that `(x, y, *, *)` exists for all `y in [C, D]` and `x` for which `(*, *, x, *)` is in `transitions`.

The second one of those would be my preferred approach.